File: proj/compiler/test_programs/pacman.py
"""pacman.py

Renders a 2D version of pacman (meant to be run with Python 3.x)
"""

import math
import numpy as np
import random
import sys; sys.path += ['../../compiler']
import util
import time
try:
    import skimage
    import skimage.color
    import skimage.io
    has_skimage = True
except ImportError:
    has_skimage = False         # For PyPy comparisons

use_gui = False

if use_gui:
    import pygame
    import time

#transform(TypeSpecialize(checks=False))
def draw_circle(x, y, r, img, color_r, color_g, color_b):
    """Draws a Pacman to img

    Args:
        x, int, x location in img
        y, int, y location in img
        r, int, radius of circle
        img, np.array, 3D color image matrix
        color_r, int, red channel of color
        color_g, int, green channel of color
        color_b, int, blue channel of color
    """

    color = (color_r, color_g, color_b)

    y_start = int(max(0, y - r))
    y_stop = int(min(y + r, img.shape[0] - 1))

    for y_i in range(y_start, y_stop):
        x_start = int(x - math.sqrt(r**2 - (y - y_i)**2))
        x_stop = int(x + math.sqrt(r**2 - (y - y_i)**2))

        for x_i in range(x_start, x_stop):
            img[x_i, y_i] = color

#transform(TypeSpecialize(checks=False))
def draw_pacman_right(x, y, r, img, ma, color_r, color_g, color_b):
    """Draws a Pacman to img

    Args:
        x, int, x location in img (center of pacman)
        y, int, y location in img (center of pacman)
        r, int, radius of pacman
        img, np.array, 3D color image matrix
        ma, float, angle mouth is open at ("mouth angle")
        color_r, float, red channel of color
        color_g, float, green channel of color
        color_b, float, blue channel of color
    """

    color = (color_r, color_g, color_b)

    y_start = int(max(0, y - r))
    y_stop = int(min(y + r, img.shape[0] - 1))

    for y_i in range(y_start, y_stop):
        x_start = int(x - math.sqrt(r**2 - (y - y_i)**2))
        x_stop = int(x + math.sqrt(r**2 - (y - y_i)**2))

        # top half of mouth:
        if y_i > y - float(r) * math.sin(ma) and y_i <= y:
            r_mouth = float(y - y_i) / math.sin(ma)
            x_stop = int(x + r_mouth * math.cos(ma))

        # bottom half of mouth:
        elif y_i < y + float(r) * math.sin(ma) and y_i > y:
            r_mouth = float(y_i - y) / math.sin(ma)
            x_stop = int(x + r_mouth * math.cos(ma))

        for x_i in range(x_start, x_stop):
            img[x_i, y_i] = color

    # draw the eye:
    draw_circle(x, y - int(r / 2), int(r / 10.), img, 0, 0, 0)

#transform(TypeSpecialize(checks=False))
def draw_rect(x, y, w, h, img, color_r, color_g, color_b):
    """Draws a rectangle to img
    
    Args:
        x, int, x location in img
        y, int, y location in img
        w, int, width
        h, int, height
        img, np.array, 3D color image matrix
        color_r, int, red channel of color
        color_g, int, green channel of color
        color_b, int, blue channel of color
    """

    color = (color_r, color_g, color_b)

    y_start = int(max(0, y))
    y_stop = int(min(y + h, img.shape[0] - 1))

    x_start = int(max(0, x))
    x_stop = int(min(x + w, img.shape[1] - 1))

    for y_i in range(y_start, y_stop):
        for x_i in range(x_start, x_stop):
            img[y_i, x_i, 0] = color_r
            img[y_i, x_i, 1] = color_g
            img[y_i, x_i, 2] = color_b

#transform(TypeSpecialize(checks=False))
def draw_ghost(x, y, r, img, color_r, color_g, color_b, tf, blink):
    """Draws a ghost to img
    
    Args:
        x, int, x location in img
        y, int, y location in img
        r, int, radius of the ghosts's head
        img, np.array, 3D color image matrix
        color_r, int, red channel of color
        color_g, int, green channel of color
        color_b, int, blue channel of color
        tf, float, how much of ghost is not tentacles
        blink, boolean, whether or not the ghost should be blinking
    """

    color = (color_r, color_g, color_b)

    y_start = int(max(0, y - r))
    y_stop = int(min(y + r, img.shape[1] - 1))

    for y_i in range(y_start, y_stop):
        x_start = int(x - math.sqrt(r**2 - (y - y_i)**2))
        x_stop = int(x + math.sqrt(r**2 - (y - y_i)**2))

        # bottom half of ghost:
        if y_i > y:
            x_start = int(max(0, x - r))
            x_stop = int(min(x + r, img.shape[0] - 1))

        # print(x_start, x_stop, y_start, y_stop)

        for x_i in range(x_start, x_stop):
            if y_i <= y + tf * r:
                img[x_i, y_i] = color
            else:
                if x_i < x - r * 5/7. or (x_i > x - r * 3/7. and x_i < x - r * 1/7.) or (x_i > x + r * 1/7. and x_i < x + r * 3/7.) or (x_i > x + r * 5/7.):
                    img[x_i, y_i] = color 

    # draw the eye:
    if not blink:
        draw_circle(x - int(r / 4), 
                    y - int(r / 2), 
                    int(r / 5.), 
                    img,
                    1, 1, 1)
        draw_circle(x + int(r / 4), 
                    y - int(r / 2), 
                    int(r / 5.), 
                    img, 
                    1, 1, 1)
        draw_circle(x - int(r / 8.), 
                    y - int(r / 2), 
                    int(r / 9.), 
                    img, 
                    0, 0, 1)
        draw_circle(x + int(r * 3 / 8.), 
                    y - int(r / 2), 
                    int(r / 9.), 
                    img, 
                    0, 0, 1)
    else:
        draw_rect(y - int(r / 2), 
                  x - int(r / 4),
                  r / 8., 
                  r / 4., 
                  img, 
                  0, 0, 0)
        draw_rect(y - int(r / 2), 
                  x + int(r / 4),
                  r / 8., 
                  r / 4., 
                  img,
                  0, 0, 0)

#transform(TypeSpecialize(checks=False))
def render_1_frame_no_write(img):
    """Renders a single frame, for testing purposes
    """

    # clear screen
    T0 = time.time()
    draw_rect(0, 0, img.shape[0], img.shape[1], img, 0, 0, 0)
    T1 = time.time()
    
    # draw pacman:
    draw_pacman_right(100, 100, 50, img, math.pi/4., 1, 1, 0)
    T2 = time.time()
    
    # draw a dot:
    draw_circle(200, 100, 15, img, 1, 1, 1)
    draw_circle(275, 100, 15, img, 1, 1, 1)
    draw_circle(350, 100, 15, img, 1, 1, 1)
    T3 = time.time()
    
    # draw a ghost:
    draw_ghost(500, 100, 50, img, 0, 1, 1, 0.75, False)
    T4 = time.time()

    print(T4-T0, T1-T0, T2-T1, T3-T2, T4-T3)

def test(n=None):
    """Default unit tests which compare any optimized output with the original Python output.
    """
    img = np.zeros((600, 200, 3))
    return util.test_image_pipeline(
        image_func=render_1_frame_no_write,
        input_imgL=(img,),
        n=None,
        ground_truth_output=None,
        use_output_img=False,
        name="pacman.render_1_frame_no_write python",
        imdiff_ignore_last_element=True)

if __name__ == "__main__":
    test()

File: proj/apps/pacman/c/pacman.c
#include "c_util.h"
#include <stdio.h>
#include <omp.h>
#include <math.h>

float max(float a, float b)
{
    return a > b ? a : b;
}

float min(float a, float b)
{
    return a < b ? a : b;
}

float clip(float input, float lower_bound, float upper_bound)
{
    return min(max(input, lower_bound), upper_bound);
}

void draw_circle(int x, int y, int r, FLOAT3_IMG *img, float color_r, float color_g, float color_b)
{
    int y_start = max(0, y - r);
    int y_stop = min(y + r, img->cols - 1);
    int x_i, y_i, x_start, x_stop;
    
    #pragma omp parallel for private(x_i, x_start, x_stop)
    for (y_i = y_start; y_i < y_stop; y_i++)
    {
	x_start = (int)(x - sqrt(r * r - (y - y_i) * (y - y_i)));
	x_stop = (int)(x + sqrt(r * r - (y - y_i) * (y - y_i)));
	
	for (x_i = x_start; x_i < x_stop; x_i++)
	{
	    img->data[y_i][x_i][0] = color_r;
	    img->data[y_i][x_i][1] = color_g;
	    img->data[y_i][x_i][2] = color_b;
	}
    }
}

void draw_pacman_right(int x, int y, int r, FLOAT3_IMG *img, float ma, float color_r, float color_g, float color_b)
{
    int y_start = max(0, y - r);
    int y_stop = min(y + r, img->cols - 1);
    int x_i, y_i, x_start, x_stop;
    float r_mouth;
    
    #pragma omp parallel for private(x_i, x_start, x_stop, r_mouth)
    for (y_i = y_start; y_i < y_stop; y_i++)
    {
	x_start = (int)(x - sqrt(r * r - (y - y_i) * (y - y_i)));
	x_stop = (int)(x + sqrt(r * r - (y - y_i) * (y - y_i)));
	
	if (y_i > (float)y - (float)r * sin(ma) && y_i <= y)
	{
	    r_mouth = (float)(y - y_i) / sin(ma);
	    x_stop = (int)(x + r_mouth * cos(ma));
	}
	else if (y_i < (float)y + (float)r * sin(ma) && y_i > y)
	{
	    r_mouth = (float)(y_i - y) / sin(ma);
	    x_stop = (int)(x + r_mouth * cos(ma));
	}
	
	for (x_i = x_start; x_i < x_stop; x_i++)
	{
	  img->data[y_i][x_i][0] = color_r;
	  img->data[y_i][x_i][1] = color_g;
	  img->data[y_i][x_i][2] = color_b;
	}
    }
    
    draw_circle(x, y - r / 2, r / 10, img, 0.0f, 0.0f, 0.0f);
}
    
void draw_rect(int x, int y, int w, int h, FLOAT3_IMG *img, float color_r, float color_g, float color_b)
{
    int y_start = max(0, y);
    int y_stop = min(y + h, img->cols - 1);
    
    int x_start = max(0, x);
    int x_stop = min(x + w, img->rows - 1);
    
    int x_i, y_i;
    
    #pragma omp parallel for private(x_i)
    for (y_i = y_start; y_i < y_stop; y_i++)
    {
	for (x_i = x_start; x_i < x_stop; x_i++)
	{
	  img->data[x_i][y_i][0] = color_r;
	  img->data[x_i][y_i][1] = color_g;
	  img->data[x_i][y_i][2] = color_b;
	}
    }
}

void draw_ghost(int x, int y, int r, FLOAT3_IMG *img, float color_r, float color_g, float color_b, float tf, int blink)
{
    int y_start = max(0, y - r);
    int y_stop = min(y + r, img->rows - 1);
    
    int x_i, y_i, x_start, x_stop;
    
    #pragma omp parallel for private(x_i, x_start, x_stop)
    for (y_i = y_start; y_i < y_stop; y_i++)
    {
	x_start = (int)(x - sqrt(r * r - (y - y_i) * (y - y_i)));
	x_stop = (int)(x + sqrt(r * r - (y - y_i) * (y - y_i)));
	
	if (y_i > y)
	{
	    x_start = max(0, x - r);
	    x_stop = min(x + r, img->cols - 1);
	}
	
	for (x_i = x_start; x_i < x_stop; x_i++)
	{
	    if (y_i <= y + tf * r)
	    {
		img->data[y_i][x_i][0] = color_r;
		img->data[y_i][x_i][1] = color_g;
		img->data[y_i][x_i][2] = color_b;
	    }
	    else
	    {
		if (x_i < x - r * (5.0f / 7.0f) || (x_i > x - r * (3.0f / 7.0f) && x_i < x - r * (1.0f / 7.0f)) || (x_i > x + r * 1.0f / 7.0f && x_i < x + r * (3.0f / 7.0f)) || x_i > x + r * (5.0f / 7.0f))
		{
		    img->data[y_i][x_i][0] = color_r;
		    img->data[y_i][x_i][1] = color_g;
		    img->data[y_i][x_i][2] = color_b;
		}
	    }
	}
    }
    
    if (!blink)
    {
	draw_circle(x - r / 4, y - r / 2, r / 5, img, 1.0f, 1.0f, 1.0f);
	draw_circle(x + r / 4, y - r / 2, r / 5, img, 1.0f, 1.0f, 1.0f);
	draw_circle(x - r / 8, y - r / 2, r / 9, img, 0.0f, 0.0f, 1.0f);
	draw_circle(x + 3 * r / 8, y - r / 2, r / 9, img, 0.0f, 0.0f, 1.0f);
    }
    else
    {
	draw_rect(y - r / 2, x - r / 4, r / 8, r / 4, img, 0.0f, 0.0f, 0.0f);
	draw_rect(y - r / 2, x + r / 4, r / 8, r / 4, img, 0.0f, 0.0f, 0.0f);
    }
}

void render_1_frame_no_write(FLOAT3_IMG *input_img, FLOAT3_IMG *output_img, float time)
{
    int ma = (int)time % 16;
    
    if (ma > 7)
	ma = 8 - (ma - 8);
    
    draw_rect(0, 0, output_img->cols, output_img->rows, output_img, 0.0f, 0.0f, 0.0f);
    
    draw_pacman_right(100, 100, 50, output_img, (float)ma * M_PI / 32.0, 1.0f, 1.0f, 0.0f);
    
    draw_circle(200, 100, 15, output_img, 1.0f, 1.0f, 1.0f);
    draw_circle(275, 100, 15, output_img, 1.0f, 1.0f, 1.0f);
    draw_circle(350, 100, 15, output_img, 1.0f, 1.0f, 1.0f);
    
    draw_ghost(500, 100, 50, output_img, 0.0f, 1.0f, 1.0f, 0.75f, ma == 7);
}

int main(int argc, char **argv)
{
    FLOAT3_IMG *input_img = zeros_rgb(200, 600);
    FLOAT3_IMG *output_img = zeros_rgb(200, 600);
    
    float time;
    double d1, d2;
    
    if (argc > 3)
    {
	printf("Usage: ./a.out or ./a.out out.png or ./a.out out.png time\n");
	return 0;
    }
    
    if (argc == 3)
    {
	time = atof(argv[2]);
	d1 = wall_time();
	render_1_frame_no_write(input_img, output_img, time);
	d2 = wall_time();
    }
    else
    {
	time = 3.0f;
	d1 = wall_time();
	render_1_frame_no_write(input_img, output_img, time);
	d2 = wall_time();
    }
    
    printf("%f\n", d2 - d1);
    
    if (argc == 1)
	write_img_rgb(output_img, "test.png");
    else
	write_img_rgb(output_img, argv[1]);
    
    return 0;
}
    
    
